### PR TITLE
Publish quality attack flow diagrams.

### LINKED ISSUE #42: Publishing quality attack flow diagrams
Related to https://github.com/center-for-threat-informed-defense/attack-flow/issues/40, how does the community use the attack flow JSON files to publish quality diagrams, such as this one?

Are you able to share what tooling was used [to generate this]?
Edit (opening question is still valid): Found the answer. They were manually(?) generated drawio diagrams (e.g. https://github.com/center-for-threat-informed-defense/attack-flow/blob/main/data/action-object-tesla.drawio).

image

### PR DESCRIPTION
**Pull request by curious-attempt-bunny**
Fixes #42 

## What Changed

1. Tidied up graphviz script output. Now it aligns better with the style of the manually created drawio diagrams.

### Before

image

### After

image

## Limitations

1. Layout! (it's graphviz)
2. Text exceeding the fixed sized nodes.

### CODE REVIEW

File: src/attack_flow/graphviz.py [hunk 1]
```diff
@@ -2,11 +2,39 @@ def convert(attack_flow):
...
-    graph.extend([f'  "{rel["source"]}" -> "{rel["target"]}"' for rel in attack_flow['relationships']])
+    graph.extend([f'  "{rel["source"]}" -> "{rel["target"]}" [label="{"provides" if rel["type"].endswith("#state-change") else "requires"}"]' for rel in attack_flow['relationships'] if '#state' in rel["type"]])
+    graph.append("")
+    graph.extend([f'  "{data_property["source"]}-{data_property["target"]}-state" [shape=box,style="rounded,filled",fillcolor=lightgreen,label="{align_node_label(data_property["target"])}",fixedsize=true,width=2,height=1]' for data_property in attack_flow["data_properties"] if data_property["type"].endswith("#state")])
+    graph.append("")
+    graph.extend([f'  "{data_property["source"]}-{data_property["target"]}-state" -> "{data_property["source"]}" [dir=none,style=dashed]' for data_property in attack_flow["data_properties"] if data_property["type"].endswith("#state")])
     graph.append("}")
     return "\n".join(graph)
+
+def align_node_label(label: str, width=20) -> str:
```
Comment (mehaase) [inline src/attack_flow/graphviz.py:25]: Can you add a docstring to this method that explains what it does?

**Review by mehaase — CHANGES_REQUESTED** [review]
Thanks for making this PR. The output looks great. I have one comment on the diff, and can you also please fix the [build errors](https://github.com/center-for-threat-informed-defense/attack-flow/runs/6489201008?check_suite_focus=true)? The build is failing because:

1. The graphviz test needs to be updated (run the test suite locally with `pytest --cov=src/ --cov-report term-missing`)
2. Some flake8 formatting issues, mainly long lines (>120 chars) in graphviz.py. I suggest unrolling some of the list comprehensions but I'll leave that up to you.

**Comment by curious-attempt-bunny** [comment]
@mehaase I addressed the feedback. Although somehow there are now 116 Security Hotspots according to SonarCloud - but I'm unable to view the report.

I also added documentation for running the tests, added the flake checks to the makefile, duplicated the flake8 configuration so that it's picked up out of the box by the MS Visual Studio plugin, and

File: src/attack_flow/graphviz.py [hunk 2]
```diff
@@ -1,12 +1,93 @@
+GRAPHVIZ_NODE_STYLE_COMMON = ''
```
Comment (mehaase) [inline src/attack_flow/graphviz.py:1]: Is this used anywhere? If not, please delete.

File: src/attack_flow/graphviz.py [hunk 3]
```diff
@@ -1,12 +1,93 @@
...
+        source = asset["id"]
+        attributes = f'fillcolor=lightblue1{label}'
+        graph.append(f'  "{source}" [{attributes}]')
+    graph.append("")
+
+    # action <-> asset arrows
+    for rel in attack_flow['relationships']:
+        if '#state' in rel.get("type", ""):
+            if rel["type"].endswith("#state-change"):
+                labelText = "provides"
```
Comment (mehaase) [inline src/attack_flow/graphviz.py:41]: Change to `label_text`.

**Review by mehaase — CHANGES_REQUESTED** [review]

**Comment by curious-attempt-bunny** [comment]
@mehasse Cleaned up. Thanks for your time reviewing these changes.

**Review by mehaase — APPROVED** [review]

**Comment by mehaase** [comment]
Thank you for the contribution @curious-attempt-bunny! This will definitely be valuable to the community at large.

**Pull request merged**